designs/0035-pre-model-gqs.md

```diff
@@ -0,0 +1,840 @@
+  double sens;
+  double spec;
+};
+```
+
+Here, the `latent_data_base_crtp` will extend `latent_data_base`,
+following the pattern used for the model class.
+
+### Constructing the latent data generically
+
+Just like the model class, this will require a generic factory method,
+though in this case, no arguments are required as it will be used like
+a simple `struct`.
+
+```cpp
+latent_data_base& new_latent_data();
+```
+
+This must then get passed where necessary and cast into the relevant
+type in the class definition, as shown in the next section.
+
+
+### Generating the latent data
+
+An additional method in the model class will be required to populate 
+the latent data. 
+
+```cpp
+template <typename RNG> inline void
+set_latent_data(
+    const VectorXd& params_r, 
+    latent_data_base& ldb, 
+    RNG& base_rng, 
+    std::ostream* pstream=nullptr) const {
+  latent_data& latent_data__ = static_cast<latent_data&>(ldb);
+  ... code generated to populate latent_data__ ...
+}
+```
+
+Because it does not use anything specific to the specific model class,
+this can be implemented inthe `latent_data_base_crtp` class using the
+CRTP.
+
+```
+template <class Derived>
+class latent_data_base_crtp<Derived> {
+template <typename RNG> inline void
+set_latent_data(
+    const VectorXd& params_r, 
+    latent_data_base& ldb, 
+    RNG& base_rng, 
+    std::ostream* pstream=nullptr) const {
+	static_cast<Derived&>(ldb).set_latent_data(params_r, ldb,
+                                               base_rng, pstream);
+};
+```
+
+The concrete `latent_data` class implemented along with the model
+class will be responsible for holding the memory for any data needed
+in latent data, such as `Eigen::Matrix` or `std::vector`
+instances. By reusing the same `latent_data` variable (per thread),
+variables can be reset rather than allocating and freeing each
+evaluation.
+
+### The `latent_data_base` class
+
+The `latent_data_base` class will only know about how to 
+write out variables and variable names.
+
+```cpp
+class latent_data_base {
+  int iteration_number_;
+  int get_iteration_number() {
+    return iteration_number_;
+  }
+  void set_iteration_number(int n) {
+    iteration_number_ = n;
+	}
+  virtual void num_variables() const = 0;
+  virtual void write_variable_names(std::vector<string>& names) const = 0;
+  virtual void write_variable_values(std::vector<double>& values) const = 0;
+};
+```
+
+The CRTP class will impelement the three virtual functions using the
+CRTP in the usual way.  
+This uses inheritance from the base class, but the cost is relatively
+low in that it is only done once per iteration.  This follows the way
+Stan's `model_base` class is designed.
+
+
+### Generalized `log_prob` method
+
+The existing `log_prob` method must be updated to maintain backward
+compatibility: 
+
+```cpp
+template <bool propto__, bool jacobian__, typename T_> inline T_
+log_prob(Eigen::Matrix<T_,-1,1>& params_r, 
```

Comment:
These two signatures contradict the above (`auto log_prob(stan::math::allocator& alloc, ...) {`)

We may want to be somewhat nonspecific in the design doc and just say "log_prob will need an additional argument" without specifying right now what the type of that argument will be